rules/sql-server-cpu-pressure/rule.md

```diff
@@ -0,0 +1,126 @@
+related: 
+  - identify-sql-performance-sql-server
+  - identify-sql-performance-azure
+  - sql-server-memory-pressure
+  - sql-server-io-pressure
+---
+So you've identified that your SQL Server is under CPU pressure. What can you do about it?
+
+<!--endintro-->
+
+Here's a simple in-depth presentation covering things that can help reduce CPU pressure.
+
+1. Should you throw hardware at the problem
+2. Update the index statistics
+3. Identify high CPU queries
+4. Identify poor queries (reading too much data either columns or rows)
+5. Identify missing indexes
+
+## Add more CPUs
+
+In many situations, the problem is poorly specifying the hardware configuration for your SQL Server. It's worth thinking about whether increasing the server specifications is the easiest solution, or whether optimising the applications are a better choice.
+
+## Update index statistics
+
+This can be achieved by
+
+``` sql
+exec sp_updatestats
+```
+
+Be aware that this can be expensive and as such using it all the time to avoid problems is not a great solution. But as a one off to verify whether your problem is a statistics issue, this is perfect.
+
+## Identifying high CPU queries
+
+The following sql identifies high CPU queries running right now.
+
+```sql
+SELECT TOP 10 s.session_id,
+           r.status,
+           r.cpu_time,
+           r.logical_reads,
+           r.reads,
+           r.writes,
+           r.total_elapsed_time / (1000 * 60) 'Elaps M',
+           SUBSTRING(st.TEXT, (r.statement_start_offset / 2) + 1,
+           ((CASE r.statement_end_offset
+                WHEN -1 THEN DATALENGTH(st.TEXT)
+                ELSE r.statement_end_offset
+            END - r.statement_start_offset) / 2) + 1) AS statement_text,
+           COALESCE(QUOTENAME(DB_NAME(st.dbid)) + N'.' + QUOTENAME(OBJECT_SCHEMA_NAME(st.objectid, st.dbid)) 
+           + N'.' + QUOTENAME(OBJECT_NAME(st.objectid, st.dbid)), '') AS command_text,
+           r.command,
+           s.login_name,
+           s.host_name,
+           s.program_name,
+           s.last_request_end_time,
+           s.login_time,
+           r.open_transaction_count
+FROM sys.dm_exec_sessions AS s
+JOIN sys.dm_exec_requests AS r ON r.session_id = s.session_id CROSS APPLY sys.Dm_exec_sql_text(r.sql_handle) AS st
+WHERE r.session_id != @@SPID
+ORDER BY r.cpu_time DESC
+```
+
+If queries aren't driving the CPU currently, try the following query.
+
+``` sql
+SELECT TOP 10 st.text AS batch_text,
+    SUBSTRING(st.TEXT, (qs.statement_start_offset / 2) + 1, ((CASE qs.statement_end_offset WHEN - 1 THEN DATALENGTH(st.TEXT) ELSE qs.statement_end_offset END - qs.statement_start_offset) / 2) + 1) AS statement_text,
+    (qs.total_worker_time / 1000) / qs.execution_count AS avg_cpu_time_ms,
+    (qs.total_elapsed_time / 1000) / qs.execution_count AS avg_elapsed_time_ms,
+    qs.total_logical_reads / qs.execution_count AS avg_logical_reads,
+    (qs.total_worker_time / 1000) AS cumulative_cpu_time_all_executions_ms,
```

Comment:
**nitpick**: Formatting consistency - are the parentheses required (when they weren't on the line above)?

Reply:
Fixed